extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -0,0 +1,291 @@
+import {timeStrToMillis} from '../../../extensions/amp-story/1.0/utils';
+import {CommonSignals} from '../../../src/common-signals';
+import {whenUpgradedToCustomElement} from '../../../src/dom';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {userAssert} from '../../../src/log';
+import {Services} from '../../../src/services';
+import {Matrix, Renderer} from '../../../third_party/zuho/zuho';
+
+
+/**
+ * Internal helper class representing a camera orientation (POV) in polar
+ * coordinates.
+ */
+class CameraOrientation {
+  /**
+   * @param {number} theta
+   * @param {number} phi
+   * @param {number} scale
+   */
+  constructor(theta, phi, scale) {
+    this.theta = theta;
+    this.phi = phi;
+    this.scale = scale;
+  }
+
+  /**
+   * @param {number} heading
+   * @param {number} pitch
+   * @param {number} zoom
+   * @return {!CameraOrientation}
+   */
+  static fromDegrees(heading, pitch, zoom) {
+    const deg2rad = (deg) => deg * Math.PI / 180;
+    return new CameraOrientation(
+        deg2rad(-pitch - 90), deg2rad(heading), 1 / zoom);
+  }
+
+  /**
+   * @return {!Float32Array}
+   */
+  get rotation() {
+    return Matrix.mul(
+        3, Matrix.rotation(3, 1, 2, this.theta),
+        Matrix.rotation(3, 0, 1, this.phi));
+  }
+}
+
+
+/**
+ * Internal helper class generating a sequence of frame by frame orientations
+ * from a set of 2 or more CameraOrientation target points and an animation
+ * duration (approximate duration in ms based on 60FPS refresh rate).
+ */
+class CameraAnimation {
+  /**
+   * @param {number} durationMs
+   * @param {!Array<!CameraOrientation>} orientations
+   */
+  constructor(durationMs, orientations) {
+    this.maxFrame = 60 / 1000 * durationMs;
+    this.orientations = orientations;
+    this.currentHeadingIndex = 0;
+    this.currentFrame = 0;
+  }
+
+  /**
+   * @param {number} x
+   * @return {number}
+   * @private
+   */
+  easeInOutQuad_(x) {
+    return x < 0.5 ? 2 * x * x : 1 - Math.pow(-2 * x + 2, 2) / 2;
+  }
+
+  /**
+   * Returns the CameraOrientation for the next frame, or null if the animation
+   * is over.
+   * @return {?CameraOrientation}
+   */
+  getNextOrientation() {
+    if (this.currentHeadingIndex < 0 ||
+        this.currentFrame == this.maxFrame - 1) {
+      // Animation ended.
+      return null;
+    }
+    this.currentFrame++;
+    const framesPerSection = this.maxFrame / (this.orientations.length - 1);
+    const lastFrameOfCurrentSection =
+        (this.currentHeadingIndex + 1) * framesPerSection;
+    if (this.currentFrame >= lastFrameOfCurrentSection) {
+      this.currentHeadingIndex++;
+      if (this.currentHeadingIndex == this.orientations.length) {
+        // End of animation.
+        this.currentHeadingIndex = -1;
+        return null;
+      } else {
+        return this.orientations[this.currentHeadingIndex];
+      }
+    }
+    const easing = this.easeInOutQuad_(
+        this.currentFrame % framesPerSection / framesPerSection);
```

Comment:
Why not `this.currentFrame / framesPerSection`? 
Do we need the `%`?

Reply:
That's because CameraAnimation supports an unlimited number of stops (sections) and the easing is reset between each. Say, if you provide 3 orientations (Start, Middle, End) then the animation will split the duration in 2 and then apply easing from Start to Middle, then again from Middle to End - so you have 2 sections. Hence the % that resets the easing. In practice currently we only provide start and end and always have 1 section so this could be removed. I don't think it's adding a lot of overhead though.

Reply:
Okay nice. Thank you for clarifying the logic!